knora/dsplib/utils/xml_upload.py

```diff
@@ -843,65 +769,147 @@ def xml_upload(input_file: str, server: str, user: str, password: str, imgdir: s
     proj_context = ProjectContext(con=con)
+    sipi_server = Sipi(sipi, con.get_token())
 
-    resources: list[XMLResource] = []
-    permissions: dict[str, XmlPermission] = {}
-
-    # parse the XML file containing the data
+    # parse the XML file
     tree = etree.parse(input_file)
-
-    # Iterate through all XML elements
     for elem in tree.getiterator():
-        # Skip comments and processing instructions,
-        # because they do not have names
-        if not (
-            isinstance(elem, etree._Comment)
-            or isinstance(elem, etree._ProcessingInstruction)
-        ):
-            # Remove a namespace URI in the element's name
-            elem.tag = etree.QName(elem).localname
-
-    # Remove unused namespace declarations
-    etree.cleanup_namespaces(tree)
+        if not (isinstance(elem, etree._Comment) or isinstance(elem, etree._ProcessingInstruction)):
+            elem.tag = etree.QName(elem).localname  # remove namespace URI in the element's name
+    etree.cleanup_namespaces(tree)  # remove unused namespace declarations
 
     knora = tree.getroot()
     default_ontology = knora.attrib['default-ontology']
     shortcode = knora.attrib['shortcode']
 
+    resources: list[XMLResource] = []
+    permissions: dict[str, XmlPermission] = {}
     for child in knora:
-        # get all permissions
         if child.tag == "permissions":
             permission = XmlPermission(child, proj_context)
             permissions[permission.id] = permission
-        # get all resources
         elif child.tag == "resource":
             resources.append(XMLResource(child, default_ontology))
 
+    # get the project information and project ontology from the server
+    project = ResourceInstanceFactory(con, shortcode)
+    permissions_lookup: dict[str, Permissions] = {s: perm.get_permission_instance() for s, perm in permissions.items()}
+    resclass_name_2_type: dict[str, type] = {s: project.get_resclass_type(s) for s in project.get_resclass_names()}
+
     # temporarily remove circular references, but only if not an incremental upload
     if not incremental:
         resources, stashed_xml_texts, stashed_resptr_props = remove_circular_references(resources, verbose)
     else:
         stashed_xml_texts = dict()
         stashed_resptr_props = dict()
 
-    sipi_server = Sipi(sipi, con.get_token())
+    id2iri_mapping: dict[str, str] = {}
+    failed_uploads: list[str] = []
 
-    # get the project information and project ontology from the server
-    project = ResourceInstanceFactory(con, shortcode)
+    try:
+        id2iri_mapping, failed_uploads = upload_resources(verbose, resources, imgdir, sipi_server, permissions_lookup,
+                                                          resclass_name_2_type, id2iri_mapping, con, failed_uploads)
+    except BaseException as err:
+        handle_upload_error(err, input_file, id2iri_mapping, failed_uploads, stashed_xml_texts, stashed_resptr_props)
+        exit(1)
```

Comment:
I think here you should do `raise err` rather than `exit(1)`, otherwise the information contained by the exception is lost to the user. This is essentially "hiding" an exception, which you should only do when you have successfully handled it.

Reply:
handle_upload_error() will now reraise the exception, instead of `exit(1)`